### PR TITLE
JSpecify mode: initial support for generic methods (with explicit type arguments at calls)

### PR DESCRIPTION
**Pull request by haewiful**
Thank you for contributing to NullAway!

Please note that once you click "Create Pull Request" you will be asked to sign our [Uber Contributor License Agreement](https://cla-assistant.io/uber/NullAway) via [CLA assistant](https://cla-assistant.io/).

Before pressing the "Create Pull Request" button, please provide the following:

  - [ ] Added support for generic identity functions with explicitly-typed nullness.

  - [ ] Related to the issue #1035.

  - [ ] void genericNonNullIdentityFunction() and void genericNullAllowingIdentityFunction() unit tests are added to GenericMethodTests.java file.

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 1]
```diff
@@ -145,6 +148,73 @@ private static boolean[] getTypeParamsWithNullableUpperBound(
     return result;
   }
 
+  public static void checkInstantiationForGenericMethodCalls(
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:151]: We should add Javadoc on this method

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 2]
```diff
@@ -145,6 +148,73 @@ private static boolean[] getTypeParamsWithNullableUpperBound(
...
+            break;
+          }
+        }
+      }
+    }
+    Symbol.MethodSymbol methodSymbol = ASTHelpers.getSymbol(methodTree);
+
+    // base type that is being instantiated
+    Type baseType = methodSymbol.asType();
+    List<Type> baseTypeArgs = baseType.getTypeArguments();
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:180]: Let's rename:
```suggestion
    List<Type> baseTypeVariables = baseType.getTypeArguments();
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 3]
```diff
@@ -715,6 +785,40 @@ public static Nullness getGenericReturnNullnessAtInvocation(
       MethodInvocationTree tree,
       VisitorState state,
       Config config) {
+
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:788]: We need to update the Javadoc of `getGenericReturnNullnessAtInvocation` to indicate it also handles generic methods, in addition to instance methods of generic classes (what it handled before)

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 4]
```diff
@@ -715,6 +785,40 @@ public static Nullness getGenericReturnNullnessAtInvocation(
...
+    List<? extends Tree> typeArgumentTrees = tree.getTypeArguments();
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:789]: Does this new logic only apply to generic methods?  If so, can we only run it for that case, by checking if the invoked method has any type variables, `!invokedMethodSymbol.getTypeParameters().isEmpty()`?

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 5]
```diff
@@ -715,6 +785,40 @@ public static Nullness getGenericReturnNullnessAtInvocation(
...
+    } else {
+      // If it's not a ForAll type, handle it as a normal MethodType
+      Type.MethodType methodTypeElse = (Type.MethodType) invokedMethodSymbol.type;
+      substitutedReturnType =
+          state
+              .getTypes()
+              .subst(
+                  methodTypeElse.restype,
+                  invokedMethodSymbol.type.getTypeArguments(),
+                  explicitTypeArgs);
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:815]: I feel like if we only apply the new logic to generic methods, we will always get a `ForAll` type and won't need this `else` case

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 6]
```diff
@@ -768,6 +884,51 @@ public static Nullness getGenericParameterNullnessAtInvocation(
       MethodInvocationTree tree,
       VisitorState state,
       Config config) {
+    List<? extends Tree> typeArgumentTrees = tree.getTypeArguments();
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:887]: Same comment as above regarding only running the new logic when a generic method is being invoked

**Review by msridhar — COMMENTED** [review]
This is a great start!  I have a few comments.  Also, can we add some tests for a method like this:
```java
    public static <T, U> void twoTypeVars(T first, U second) {
    }
```
Maybe write some variants where only one of the type variables has a `@Nullable` upper bound, and some calls where the wrong type argument is made `@Nullable`.  This is just to check that our indexing logic is matching type arguments to type variables correctly.

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 7]
```diff
@@ -145,6 +148,83 @@ private static boolean[] getTypeParamsWithNullableUpperBound(
     return result;
   }
 
+  /**
+   * Checks instantiated generic arguments of generic method calls. {@code @Nullable} types are only
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:152]:
```suggestion
   * Checks instantiated generic arguments of generic method calls. {@code @Nullable} types can only be
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 8]
```diff
@@ -145,6 +148,83 @@ private static boolean[] getTypeParamsWithNullableUpperBound(
...
+    for (int i = 0; i < baseTypeVariables.size(); i++) {
+      if (nullableTypeArguments.containsKey(i)) {
+        Type typeVariable = baseTypeVariables.get(i);
+        Type upperBound = typeVariable.getUpperBound();
+        com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
+            upperBound.getAnnotationMirrors();
+        boolean hasNullableAnnotation =
+            Nullness.hasNullableAnnotation(annotationMirrors.stream(), config)
+                || handler.onOverrideTypeParameterUpperBound(baseType.tsym.toString(), i);
+        // if type argument does not have @Nullable annotation then the instantiation is invalid
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:201]:
```suggestion
        // if type variable's upper bound does not have @Nullable annotation then the instantiation is invalid
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 9]
```diff
@@ -682,7 +762,8 @@ public static Nullness getGenericMethodReturnTypeNullness(
   /**
    * Computes the nullness of the return of a generic method at an invocation, in the context of the
    * declared type of its receiver argument. If the return type is a type variable, its nullness
-   * depends on the nullability of the corresponding type parameter in the receiver's type.
+   * depends on the nullability of the corresponding type parameter in the receiver's type or the
+   * type argument of the method call.
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:766]: Eventually we should handle generic instance methods of generic types, e.g.:
```java
class Foo<T extends @Nullable Object> {
  public void <U> T test(U u) { ... }
}
...
Foo<@Nullable String> f = new Foo<@Nullable String>();
String s = f.<Integer>(3);
// Here s is @Nullable
```
But it doesn't have to be part of this pull request.  I don't think the code below will work for this case.

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 10]
```diff
@@ -768,6 +888,40 @@ public static Nullness getGenericParameterNullnessAtInvocation(
       MethodInvocationTree tree,
       VisitorState state,
       Config config) {
+    // If generic method invocation
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:891]: Similar comment to the one above about a generic instance method of a generic type

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 11]
```diff
@@ -768,6 +888,40 @@ public static Nullness getGenericParameterNullnessAtInvocation(
...
+
+      if (substitutedParamTypes != null
+          && Objects.equals(
+              getTypeNullness(substitutedParamTypes.get(paramIndex), config), Nullness.NULLABLE)) {
+        return Nullness.NULLABLE;
+      }
+    }
+
+    // problem is that it returns nullable from the declaration even though the parameter itself is
+    // nonnull
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:924]: do we still want this new comment?

**Review by msridhar — COMMENTED** [review]
Looking good!  A few more minor comments

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java (hunk 10 above)
Comment (haewiful) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:891]: Do you mean the Javadoc comment? Would you want me to erase this?

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java (hunk 9 above)
Comment (haewiful) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:766]: Should I still add the test case?

Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:764]: Sure, please add it.  If it doesn't work, and it's not easy to fix, just `@Ignore` the test and we can fix in a follow-up

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java (hunk 10 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:889]: I just meant that this code may not work in cases like https://github.com/uber/NullAway/pull/1053/files#r1805634821.  Nothing to do here for now I think, we can look in a follow up.

**Review by msridhar — COMMENTED** [review]
This mostly looks good!  Once we add a test case as discussed in the comment and see if it works, then we can land

**Comment by msridhar** [comment]
This support is incomplete (even for cases with explicit type arguments), but let's go ahead and land it and do things in pieces.  Thanks!

**Review by msridhar — APPROVED** [review]

**Pull request merged**